### DISCUSSION TITLE
AddStandardResilienceHandler: snapshot the http request message?

### DISCUSSION THREAD

**Discussion by tore-hammervoll** [dotnet/extensions#4965] [Q&A]
Looking through the implementation of `AddStandardHedgingHandler()` I see that it adds snapshotting logic to the resilience handler, to ensure each hedged attempt operate on a separate copied instance of the `HttpRequestMessage`. I can see how this is necessary to allow each hedged attempt to use a different request URI.

The `AddStandardResilienceHandler()` does not have the same snapshotting logic, which I assume is because it is not needed.

A side effect of this is that any delegating handler later in the chain will receive a unique instance of the `HttpRequestMessage` for each hedged attempt, while with the standard resilience handler they will receive the same instance again for each retry attempt.

With the standard resilience handler this can cause issues if a later delegating handler is not written to handle processing the same `HttpRequestMessage` more than once. 

Is there an expectation that a delegating handler should be able to process the same `HttpRequestMessage` more than once? If not, should the standard resilience handler be updated to also include similar snapshotting logic as the standard hedging handler?

Some context:
The retry logic of the standard resilience handler is causing issues with a delegating handler we use that adds a signature header to the request. The signature is added twice, and causes the request to fail. I could move the the signature handler before the resilience handler as workaround, but that would cause the signature to be reused triggering request replay detection.

**Comment by RussKie** [comment]
@dotnet/dotnet-r9-resilience could you help please?

**Comment by lvxiaoxin** [comment]
I got the same questions here. 

Should that always snapshot and later clone a HttpRequestMessage for hedging/retry the best practice? Today the standard hedging pipeline has the clone logic while the standard resilience pipeline doesn't.

In addition, in the clone implementation of standard hedging, it called out not to handle a StreamContent type of request body. [https://source.dot.net/#Microsoft.Extensions.Http.Resilience/Internal/RequestMessageSnapshot.cs,78](https://source.dot.net/#Microsoft.Extensions.Http.Resilience/Internal/RequestMessageSnapshot.cs,78)
Is it possible to support or not? I'm wondering, if it's a seek-able stream which can move the read position to the beginning.
